### ISSUE TITLE
`meson introspect` can hang in some cases

### ISSUE THREAD

**Issue by kcgen** [mesonbuild/meson#11763]
**Describe the bug**
When fed DOSBox Staging's meson.build file, `meson introspect` hangs.

**To Reproduce**

``` shell
cd /tmp
git clone --depth 1 https://github.com/dosbox-staging/dosbox-staging
cd dosbox-staging
meson introspect --projectinfo meson.build
```

**Expected behavior**
Printing of the project information to stdout, in JSON format.

**system parameters**
* cross build? : no
* operating system: Confirmed on Arch (latest), Ubuntu 22.04 and 23.04
* Python version: 3.11.2
* `meson --version`: 1.1.0

Thanks to @granminigun for flagging this via IRC.

**Comment by bruchar1** [comment]
The culprit is `libsdlcd_sources = files(libsdlcd_sources)` in src\libs\sdlcd\meson.build:20 .

Minimal example to reproduce:

```
project('sample', 'c')

f = ['sample.c']
f = files(f)

exe = executable('sample', f, install : true)
```

**Comment by bruchar1** [comment]
Similar problem occurs when performing a plusassignment. In `ast/interpreter.py`, `assignment` and `evaluate_plusassign` functions both replace the current assigned value by the new value, causing previous value to be lost. In the case of a `+=`, resulting introspection data would only preserve the added values of the assignment.

**Comment by bruchar1** [comment]
For example, `meson introspect --targets meson.build` on this file:

```
project('sample', 'c')

f = ['sample1.c']
f += ['sample2.c']

exe = executable('sample', f)
```

results in

```
[{"name": "sample", "id": "sample@exe", "type": "executable", "defined_in": "meson.build", "filename": ["sample.exe"], "build_by_default": true, "target_sources": [{"language": "unknown", "compiler": [], "parameters": [], "sources": ["C:\\env\\meson\\sample\\sample2.c"], "generated_sources": []}], "depends": [], "extra_files": [], "subproject": null, "installed": false}]
```

`sample1.c` is missing from introspection data.

**Comment by Volker-Weissmann** [comment]
> **Describe the bug** When fed DOSBox Staging's meson.build file, `meson introspect` hangs.
> 
> **To Reproduce**
> 
> ```shell
> cd /tmp
> git clone --depth 1 https://github.com/dosbox-staging/dosbox-staging
> cd dosbox-staging
> meson introspect --projectinfo meson.build
> ```
> 
> **Expected behavior** Printing of the project information to stdout, in JSON format.
> 
> **system parameters**
> 
>     * cross build? : no
> 
>     * operating system: Confirmed on Arch (latest), Ubuntu 22.04 and 23.04
> 
>     * Python version: 3.11.2
> 
>     * `meson --version`: 1.1.0
> 
> 
> Thanks to @GranMinigun for flagging this via IRC.

If you replace `meson.build` with the path to a build directory it works:
```
meson introspect --projectinfo builddir
```

**Comment by GranMinigun** [comment]
This is an issue with IDEs like KDevelop and Qt Creator which do it that way.

**Comment by Volker-Weissmann** [comment]
To avoid duplicate work, please be informed that I'm currently working on it, fix will (hopefully) come this week.

**Comment by eli-schwartz** [comment]
Awesome, I wasn't working on it at all beyond the analysis I documented above.

**Comment by Volker-Weissmann** [comment]
While fixing this bug I found quite a lot of limitations/bugs in the meson rewriter. I think I'm rewrite some major parts of the rewriter.

**Comment by Volker-Weissmann** [comment]
> @Volker-Weissmann did you finally fixed this bug?

Yes: It is waiting for review, then I will squash the commits and we can merge.

https://github.com/mesonbuild/meson/pull/12277
